File: shared/imessage-message.ts
/** iMessage inbound/outbound markers shared by the harness, desktop, phone, and relay. */

export const FROM_IMESSAGE_TAG = "[from iMessage]";
export const TO_IMESSAGE_TAG = "[to iMessage]";
export const IMESSAGE_INBOUND_MARKER = "IMESSAGE INBOUND";

/** Always in the bot persona so Simple-mode shared threads still know the gate. */
export const IMESSAGE_PERSONA_RULE =
  "iMessage channel: Incoming iMessage turns begin with [from iMessage] (and may sit inside an IMESSAGE INBOUND block). They are not something the owner typed in BotFleet. When a reply should go back to iMessage, start that reply with [to iMessage] on the first line, then the message. Only those tagged replies are sent to iMessage, and the tag is stripped before sending. Replies that stay in BotFleet must not use that tag.";

const FROM_RE = /^\[from iMessage\]\s*/i;
const TO_RE = /^\[to iMessage\]\s*/i;

function block(text: string, marker: string): string | undefined {
  const match = text.match(new RegExp(`\\[${marker}\\]\\n([\\s\\S]*?)\\n\\[\\/${marker}\\]`));
  return match?.[1]?.trim() || undefined;
}

export function isImessageInboundSource(source: unknown, userAgent?: string | null): boolean {
  if (typeof source === "string" && source.trim().toLowerCase() === "imessage") return true;
  return /BotFleet-Relay/i.test(userAgent ?? "");
}

export function parseImessageInbound(text: string): { body: string } | null {
  const trimmed = text.trim();
  if (!trimmed) return null;
  const inner = block(trimmed, IMESSAGE_INBOUND_MARKER);
  const raw = inner ?? trimmed;
  if (!FROM_RE.test(raw) && inner === undefined) return null;
  const body = raw.replace(FROM_RE, "").trim();
  if (!body) return null;
  return { body };
}

export interface ImessageMessageView {
  headline: string;
  subtitle: string;
  payload?: string;
  body: string;
}

/** Convert a stored iMessage inbound prompt into the smaller view shown in chat. */
export function imessageMessageView(text: string): ImessageMessageView | null {
  const parsed = parseImessageInbound(text);
  if (!parsed) return null;
  const lines = parsed.body.split("\n");
  const first = (lines[0] ?? "").trim() || "iMessage";
  const rest = lines.slice(1).join("\n").trim();
  return {
    headline: first,
    subtitle: "iMessage",
    payload: rest || undefined,
    body: parsed.body,
  };
}

/** Wrap inbound iMessage text so the model sees the tag and the UI can card it. */
export function wrapImessageInbound(text: string): string {
  const trimmed = text.trim();
  const parsed = parseImessageInbound(trimmed);
  const body = (parsed?.body ?? trimmed.replace(FROM_RE, "").trim());
  if (!body) return trimmed;
  return `[${IMESSAGE_INBOUND_MARKER}]\n${FROM_IMESSAGE_TAG}\n${body}\n[/${IMESSAGE_INBOUND_MARKER}]`;
}

/** Body of a bot reply meant for iMessage, or null when the tag is absent. */
export function stripToImessagePrefix(text: string): string | null {
  const trimmed = text.trim();
  const match = TO_RE.exec(trimmed);
  if (!match) return null;
  return trimmed.slice(match[0].length).trim();
}

/** Text the relay should send over iMessage, or null to keep the reply in-app. */
export function outboundImessageText(text: string): string | null {
  const rest = stripToImessagePrefix(text);
  if (rest === null) return null;
  return rest.trim() ? rest.trim() : null;
}

File: ios/Sources/CompanionCore/ImessageMessage.swift
// Parse a harness iMessage inbound prompt into the smaller view shown in chat.
//
// The stored message stays role=user so the model still sees it as the turn
// prompt.  The phone must not paint that as something the owner typed.
import Foundation

public struct ImessageMessageView: Equatable, Sendable {
    public var headline: String
    public var subtitle: String
    public var payload: String?
    public var body: String

    public static func parse(_ text: String?) -> ImessageMessageView? {
        guard let text, !text.isEmpty else { return nil }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let inner = block("IMESSAGE INBOUND", in: trimmed)
        let raw = inner ?? trimmed
        let hasPrefix = raw.range(
            of: #"^\[from iMessage\]\s*"#,
            options: [.regularExpression, .caseInsensitive]
        ) != nil
        if inner == nil && !hasPrefix { return nil }

        var body = raw
        if let range = body.range(
            of: #"^\[from iMessage\]\s*"#,
            options: [.regularExpression, .caseInsensitive]
        ) {
            body = String(body[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        guard !body.isEmpty else { return nil }

        let lines = body.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        let first = lines.first?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let headline = first.isEmpty ? "iMessage" : first
        let rest = lines.dropFirst().joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return ImessageMessageView(
            headline: headline,
            subtitle: "iMessage",
            payload: rest.isEmpty ? nil : rest,
            body: body
        )
    }

    /// Body of a bot reply meant for iMessage, or nil when the tag is absent.
    public static func stripToImessagePrefix(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let range = trimmed.range(
            of: #"^\[to iMessage\]\s*"#,
            options: [.regularExpression, .caseInsensitive]
        ) else { return nil }
        return String(trimmed[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func block(_ marker: String, in text: String) -> String? {
        let start = "[\(marker)]\n"
        let end = "\n[/\(marker)]"
        guard let startRange = text.range(of: start) else { return nil }
        guard let endRange = text.range(of: end, range: startRange.upperBound..<text.endIndex) else { return nil }
        let inner = text[startRange.upperBound..<endRange.lowerBound]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return inner.isEmpty ? nil : String(inner)
    }
}
